File: miscellaneous/sorting/quick-sort/20221105/quick-sort.ts
function quickSort(nums: number[], start: number, end: number): number[] {
    const pivot = end;
    let left = start;
    let right = start;

    while (right < end) {
        if (nums[right] <= nums[pivot]) {
            [nums[left], nums[right]] = [nums[right], nums[left]];
            left++;
        }

        right++;
    }

    [nums[pivot], nums[left]] = [nums[left], nums[pivot]];

    start < left - 1 && quickSort(nums, start, left - 1);
    end > left + 1 && quickSort(nums, left + 1, end);

    return nums;
}

export default quickSort;

File: miscellaneous/sorting/quick-sort/20221105/quick_sort.py
from typing import List


def quick_sort(nums: List[int], start: int, end: int) -> List[int]:
    if len(nums) <= 1:
        return nums

    pivot = end
    left = start
    right = start

    while right < end:
        if nums[right] <= nums[pivot]:
            [nums[left], nums[right]] = [nums[right], nums[left]]
            left += 1

        right += 1

    [nums[left], nums[pivot]] = [nums[pivot], nums[left]]

    left - 1 > start and quick_sort(nums, start, left - 1)
    left + 1 < end and quick_sort(nums, left + 1, end)

    return nums
